### ISSUE TITLE
Checkpoint or Training strategy of MoviNet on Charades Dataset

### ISSUE THREAD

**Issue by Wangbk-dl** [tensorflow/models#11055]
## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/blob/master/official/projects/movinet

## 2. Describe the feature you request

Could you provide the checkpoint of MoviNet on Charades dataset or training details or fine-tune details? I really need the result on Charades dataset.

**Comment by laxmareddyp** [comment]
Hi @Wangbk-dl,


As of now, pre-trained checkpoints for the Charades dataset are not available. However, you have the option to train and fine-tune your own custom dataset using the [movinet tutorial notebook](https://github.com/tensorflow/models/blob/master/official/projects/movinet/movinet_streaming_model_training_and_inference.ipynb) provided. This allows you to adapt the movinet model to your specific dataset and tasks by training it from scratch or fine-tuning on your data.

**Comment by Wangbk-dl** [comment]
> 
Thanks for your reply.
Could provide more details about your experiments on the Charades? Should I train the network without loading the pre-trained weight or just fine-tune on the Charades dataset? If fine-tune, should I only fine-tune the final classifier of the whole network? 
Really looking forward to your reply!

**Comment by laxmareddyp** [comment]
Hi @Wangbk-dl ,

It is common to use pre-trained weights from models trained on large-scale datasets and often leads to better performance, especially when the target dataset is relatively small.

Training from Scratch :This method may require a large amount of data and computational resources, and it might not perform as well as transfer learning.

Fine-tuning:This approach is commonly used and often achieves better results than training from scratch, especially when you have limited data.

If fine-tune, should I only fine-tune the final classifier of the whole network?  

Ans: For best results on the Charades dataset, fine-tuning with pre-trained weights is often the preferred approach. Start with a pre-trained model and experiment with fine-tuning various layers, focusing on the final classifier and some preceding layers, and monitor performance to find the best configuration for your specific task.

Thanks.

**Comment by Wangbk-dl** [comment]
> Hi @Wangbk-dl ,
> 
> It is common to use pre-trained weights from models trained on large-scale datasets and often leads to better performance, especially when the target dataset is relatively small.
> 
> Training from Scratch :This method may require a large amount of data and computational resources, and it might not perform as well as transfer learning.
> 
> Fine-tuning:This approach is commonly used and often achieves better results than training from scratch, especially when you have limited data.
> 
> If fine-tune, should I only fine-tune the final classifier of the whole network?
> 
> Ans: For best results on the Charades dataset, fine-tuning with pre-trained weights is often the preferred approach. Start with a pre-trained model and experiment with fine-tuning various layers, focusing on the final classifier and some preceding layers, and monitor performance to find the best configuration for your specific task.
> 
> Thanks.

Thanks so much! Best wishes!

**Comment by laxmareddyp** [comment]
Closing as this completed.

**Comment by Wangbk-dl** [comment]
Sorry for asking again.I have tried several strategies of training Movinet on Charades dataset, but the best acc only achieves 6%. I have tried load the pre-trained Movinet which has been trained on Kinetics 600 and only set the last classifier trainable. But the acc seems extremely low at first, only about 2%. Is that correct? I've also tried to train the Movinet without loading the pre-trained checkpoint. The result is also bad. I really want to know what problem I've got here. Here is my code:

`
import os.path

import tensorflow as tf
from official.projects.movinet.modeling import movinet
from official.projects.movinet.modeling import movinet_model
from charades_dataset import CharadesDataset, match_video_and_label


gpus = tf.config.experimental.list_physical_devices('GPU')
# tf.config.experimental.set_virtual_device_configuration(gpus[0], [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=4096)])
tf.config.experimental.set_memory_growth(gpus[0], True)


def build_classifier(backbone, num_class, input_shape, freeze_backbone=True):
    model = movinet_model.MovinetClassifier(
        backbone, num_classes=num_class)
    model.build([input_shape])

    if freeze_backbone:
        model.layers[0].trainable = False
        back = model.layers[1]
        for i, lay in enumerate(back.layers):
            if i < len(back.layers) - 2:
                lay.trainable = False
            else:
                lay.trainable = True
        # for layer in model.layers[1:-1]:
        #     layer.trainable = True
        model.layers[-1].trainable = True

    return model


if __name__ == '__main__':
    model_id = 'a3'
    use_positional_encoding = True

    num_class = 157

    num_epoch = 30
    batch_size = 16
    num_frames = 8
    resolution = 172

    input_shape = [batch_size, num_frames, resolution, resolution, 3]

    backbone = movinet.Movinet(
        model_id=model_id,
        causal=False,
        conv_type='2plus1d',
        se_type='2plus3d',
        activation='hard_swish',
        gating_activation='hard_sigmoid',
        use_positional_encoding=use_positional_encoding,
        use_external_states=False
    )

    model = movinet_model.MovinetClassifier(backbone, num_classes=600, output_states=True)
    model.build([batch_size, num_frames, resolution, resolution, 3])

    checkpoint_dir = '../checkpoint/movinet_a3_stream/'
    checkpoint_dir = './checkpoint/'

    checkpoint_path = tf.train.latest_checkpoint(checkpoint_dir)
    checkpoint = tf.train.Checkpoint(model=model)
    status = checkpoint.restore(checkpoint_path)
    status.assert_existing_objects_matched()

    model = build_classifier(backbone, num_class=num_class, input_shape=input_shape)
    model.build(input_shape=input_shape)

    distribution_strategy = tf.distribute.get_strategy()
    print('running on single GPU', gpus[0].name)

    data_root = '/home/wbk/code/video_understanding/Charades_v1_480/'
    label_root = '/home/wbk/code/video_understanding/Charades/Charades_v1_train.csv'
    dataset_name = 'charades'
    videos, labels = match_video_and_label(data_root, label_root)

    data_output_signature = (tf.TensorSpec(shape=(num_frames, resolution, resolution, 3), dtype=tf.float32),
                             tf.TensorSpec(shape=(num_class,), dtype=tf.float32))

    # data_output_signature = (tf.TensorSpec(shape=(num_frames, resolution, resolution, 3), dtype=tf.float32),
    #                          tf.TensorSpec(shape=(), dtype=tf.float32))

    train_videos = videos[:45000]
    train_labels = labels[:45000]

    test_videos = videos[45000:]
    test_labels = labels[45000:]

    train_dataset = CharadesDataset(train_videos, train_labels, num_frames, (resolution, resolution), training=True)
    test_dataset = CharadesDataset(test_videos, test_labels, num_frames, (resolution, resolution), training=False)

    # num_data1 = len(train_dataset)
    # print('num_training_data: {}'.format(num_data1))
    # num_data2 = len(test_dataset)
    # print('num_testing_data: {}'.format(num_data2))

    train_dataloader = tf.data.Dataset.from_generator(train_dataset, output_signature=data_output_signature)
    train_dataloader = train_dataloader.batch(batch_size=batch_size)
    # train_dataloader = train_dataloader.repeat(count=num_epoch)

    test_dataloader = tf.data.Dataset.from_generator(test_dataset, output_signature=data_output_signature)
    test_dataloader = test_dataloader.batch(batch_size=batch_size)
    # test_dataloader = test_dataloader.repeat(count=num_epoch)

    frame, lae = next(iter(train_dataloader))
    print(frame.shape, lae.shape)

    # frame, lae = next(iter(test_dataloader))
    # print(frame.shape, lae.shape)

    train_steps = len(train_dataset) // batch_size
    test_steps = len(test_dataset) // batch_size
    # train_steps = 1
    total_steps = train_steps * num_epoch

    with distribution_strategy.scope():
        model = build_classifier(backbone, num_class=num_class, input_shape=input_shape)
        model.build(input_shape=input_shape)

        loss = tf.keras.losses.CategoricalCrossentropy(
            from_logits=True,
            label_smoothing=0.1
        )

        metrics = [
            tf.keras.metrics.TopKCategoricalAccuracy(
                k=1, name='top_1', dtype=tf.float32),
            tf.keras.metrics.TopKCategoricalAccuracy(
                k=5, name='top_5', dtype=tf.float32)
        ]

        initial_lr = 0.001
        lr = tf.keras.optimizers.schedules.CosineDecay(
            initial_lr, decay_steps=train_steps
        )

        optimizer = tf.keras.optimizers.RMSprop(
            lr, rho=0.9, momentum=0.9, epsilon=1.0, clipnorm=1.0
        )
        model.compile(loss=loss, optimizer=optimizer, metrics=metrics)

    # checkpoint_dir = './checkpoint2/'
    # checkpoint_path = tf.train.latest_checkpoint(checkpoint_dir)
    # checkpoint = tf.train.Checkpoint(model=model)
    # status = checkpoint.restore(checkpoint_path)
    # status.assert_existing_objects_matched()

    checkpoint_path = 'checkpoint/cp.ckpt'
    checkpoint_dir_ = os.path.dirname(checkpoint_path)
    model_callbacks = tf.keras.callbacks.ModelCheckpoint(
        filepath=checkpoint_path,
        save_weights_only=False,
        verbose=1,
        save_best_only=True
    )

    callbacks = [
        tf.keras.callbacks.TensorBoard(),
        model_callbacks
    ]

    results = model.fit(
        train_dataloader,
        validation_data=test_dataloader,
        epochs=num_epoch,
        steps_per_epoch=train_steps,
        validation_steps=test_steps,
        callbacks=callbacks,
        validation_freq=1,
        verbose=1
    )

    model.evaluate(test_dataloader)


`
